### PR TITLE
Enable retention based on folder mod time

### PR DESCRIPTION
**Pull request by zliu41**
@ibuenros can you review?

### CODE REVIEW

File: gobblin-data-management/src/main/java/gobblin/data/management/retention/profile/SimpleDatasetProfile.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * {@link gobblin.data.management.retention.dataset.finder.DatasetFinder} for simple datasets.
+ *
+ * Simple datasets will be cleaned by the modification timestamps of the datasets that match
+ * 'gobblin.retention.dataset.pattern'.
+ */
+public class SimpleDatasetProfile extends ConfigurableGlobDatasetFinder {
+  public SimpleDatasetProfile(FileSystem fs, Properties props) throws IOException {
+    super(fs, props);
+  }
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/retention/profile/SimpleDatasetProfile.java]: You need to override the method `dataseAtPath(Path)`.

File: gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/SimpleDataset.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import gobblin.data.management.retention.version.finder.ModDateTimeDatasetVersionFinder;
+import gobblin.data.management.retention.version.finder.VersionFinder;
+
+
+/**
+ * {@link DatasetBase} for a simple dataset.
+ *
+ * Uses a {@link ModDateTimeDatasetVersionFinder} and a {@link TimeBasedRetentionPolicy}.
+ */
+public class SimpleDataset extends DatasetBase<TimestampedDatasetVersion> {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/SimpleDataset.java]: Can we call this something different from SimpleDataset? (not very informative). Maybe ModificationTimeDataset?

File: gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/ModDateTimeDatasetVersionFinder.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+/**
+ * {@link DatasetVersionFinder} for datasets based on modification timestamps.
+ */
+public class ModDateTimeDatasetVersionFinder extends DatasetVersionFinder<TimestampedDatasetVersion> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ModDateTimeDatasetVersionFinder.class);
+
+  private static final Path GLOB_PATTERN = new Path(".");
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/ModDateTimeDatasetVersionFinder.java]: Will this glob correctly? I don't understand exactly what your datasets are and what your dataset versions are, can you give an example?

Comment (zliu41) [inline gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/ModDateTimeDatasetVersionFinder.java]: Yes. suppose we have
```
/root
    /a
    /b
```
If `gobblin.retention.dataset.pattern=/root/*` then datasets will be `a` and `b`, the version is their modification time.

**Comment by zliu41** [comment]
Updated.

File: gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/ModDateTimeDatasetVersionFinder.java (hunk 3 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/ModDateTimeDatasetVersionFinder.java]: You can actually use `SingleVersionFinder` for this. It uses FileStatusDatasetVersion which does not require an additional call to `fs.filestatus`

File: gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/PathDateTimeDatasetVersionFinder.java [hunk 4]
```diff
@@ -27,19 +27,20 @@
...
 
 /**
- * {@link gobblin.data.management.retention.version.finder.DatasetVersionFinder} for timestamped datasets. Uses a datetime pattern
- * to find dataset versions and parse the {@link org.joda.time.DateTime} representing the version.
+ * {@link gobblin.data.management.retention.version.finder.DatasetVersionFinder} for datasets based on path timestamps.
+ * Uses a datetime pattern to find dataset versions from the dataset path
+ * and parse the {@link org.joda.time.DateTime} representing the version.
  */
-public class DateTimeDatasetVersionFinder extends DatasetVersionFinder<TimestampedDatasetVersion> {
+public class PathDateTimeDatasetVersionFinder extends DatasetVersionFinder<TimestampedDatasetVersion> {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/PathDateTimeDatasetVersionFinder.java]: This name is certainly better, but this class might be already used in some running projects. Can we revert it to the original name? Thanks.

**Comment by zliu41** [comment]
updated.

**Comment by ibuenros** [comment]
LGTM

**Pull request merged**